### PR TITLE
feat: add `ready()`

### PR DESCRIPTION
**Pull request by eunjae-lee**
## Summary

This PR adds `ready()` method.

## Details

It enables users to do things like the following:

```js
window.aa("ready", () => {
  console.log("The library is loaded.");
});
```

The name is inspired by jQuery's ready() and [Segment's ready()](https://segment.com/docs/connections/sources/catalog/libraries/website/javascript/#ready).

When `search-insights` is inserted like the following:

```html
<script>
var ALGOLIA_INSIGHTS_SRC = "https://cdn.jsdelivr.net/npm/search-insights@1.7.1";

!function(e,a,t,n,s,i,c){e.AlgoliaAnalyticsObject=s,e[s]=e[s]||function(){
(e[s].queue=e[s].queue||[]).push(arguments)},i=a.createElement(t),c=a.getElementsByTagName(t)[0],
i.async=1,i.src=n,c.parentNode.insertBefore(i,c)
}(window,document,"script",ALGOLIA_INSIGHTS_SRC,"aa");
</script>
```

It's hard to figure out if the library is loaded, or `window.aa` is still just an array of pending queue.

With this addition, no matter CJS or UMD, we can act on it as soon as the library is ready.

### CODE REVIEW

**Comment by eunjae-lee** [comment]
Can I get a review by the end of today? :) @tkrugg @shortcuts

File: lib/ready.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+export function ready(callback: Function) {
+  if (typeof callback === "function") {
+    callback();
+  }
```
Comment (tkrugg) [inline lib/ready.ts:4]: Shouldn't this throw or warn if it's uncorrect?

File: lib/__tests__/_processQueue.test.ts [hunk 2]
```diff
@@ -74,4 +76,20 @@ describe("processQueue", () => {
     const newPointerFunction = globalObject.aa;
     expect(oldPointerFunction).toBe(newPointerFunction);
   });
+
+  it("should executes the callback as soon as ready", () => {
```
Comment (tkrugg) [inline lib/__tests__/_processQueue.test.ts:80]:
```suggestion
  it("should execute the callback as soon as ready", () => {
```

File: lib/__tests__/ready.test.ts [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import AlgoliaAnalytics from "../insights";
+
+describe("ready()", () => {
+  it("Should execute the callback", () => {
```
Comment (tkrugg) [inline lib/__tests__/ready.test.ts:4]:
```suggestion
  it("should execute the callback", () => {
```

**Review by tkrugg — COMMENTED** [review]
It's still not clear to me what's the use case.
You're saying:

> It's hard to figure out if the library is loaded, or window.aa is still just an array of pending queue.
> With this addition, no matter CJS or UMD, we can act on it as soon as the library is ready.

but can you give me example?

**Comment by eunjae-lee** [comment]
> It's still not clear to me what's the use case.
> You're saying:
> 
> > It's hard to figure out if the library is loaded, or window.aa is still just an array of pending queue.
> > With this addition, no matter CJS or UMD, we can act on it as soon as the library is ready.
> 
> but can you give me example of a thing we can't currently do, and `ready` will make possible?

You're right. I'm closing this PR.
I thought I needed this because there are a bunch of logics in the insights middleware in InstantSearch.js in case the umd version of search-insights is asynchronously loaded, and there can be an anonymous userToken in the cookie, there can be a userToken in the aa.queue that is not consumed yet, etc.
So I wanted a way to make sure everything is ready and I can start acting on it.

However, after Youcef's question, I looked at the implementation of the insights middleware, and thanks to [the recent addition](https://github.com/algolia/search-insights.js/pull/251) in search-insights, now we can get rid of a bunch of unnecessary check-ups in the insights middleware.

```diff
diff --git a/src/middlewares/createInsightsMiddleware.ts b/src/middlewares/createInsightsMiddleware.ts
index 68f25b8f6..d13f066b2 100644
--- a/src/middlewares/createInsightsMiddleware.ts
+++ b/src/middlewares/createInsightsMiddleware.ts
@@ -49,40 +49,11 @@ export const createInsightsMiddleware: CreateInsightsMiddleware = props => {
     }
   }
 
-  const hasInsightsClient = Boolean(_insightsClient);
   const insightsClient =
     _insightsClient === null ? (noop as InsightsClient) : _insightsClient;
 
   return ({ instantSearchInstance }) => {
     const [appId, apiKey] = getAppIdAndApiKey(instantSearchInstance.client);
-    let queuedUserToken: string | undefined = undefined;
-    let userTokenBeforeInit: string | undefined = undefined;
-
-    if (Array.isArray(insightsClient.queue)) {
-      // Context: The umd build of search-insights is asynchronously loaded by the snippet.
-      //
-      // When user calls `aa('setUserToken', 'my-user-token')` before `search-insights` is loaded,
-      // ['setUserToken', 'my-user-token'] gets stored in `aa.queue`.
-      // Whenever `search-insights` is finally loaded, it will process the queue.
-      //
-      // But here's the reason why we handle it here:
-      // At this point, even though `search-insights` is not loaded yet,
-      // we still want to read the token from the queue.
-      // Otherwise, the first search call will be fired without the token.
-      [, queuedUserToken] =
-        find(
-          insightsClient.queue.slice().reverse(),
-          ([method]) => method === 'setUserToken'
-        ) || [];
-    }
-    insightsClient('getUserToken', null, (_error: any, userToken: string) => {
-      // If user has called `aa('setUserToken', 'my-user-token')` before creating
-      // the `insights` middleware, we store them temporarily and
-      // set it later on.
-      //
-      // Otherwise, the `init` call might override it with anonymous user token.
-      userTokenBeforeInit = userToken;
-    });
     insightsClient('init', { appId, apiKey, ...insightsInitParams });
 
     return {
@@ -103,21 +74,6 @@ export const createInsightsMiddleware: CreateInsightsMiddleware = props => {
 
         helper.setState(helper.state.setQueryParameter('clickAnalytics', true));
 
-        const anonymousUserToken = getInsightsAnonymousUserTokenInternal();
-        if (hasInsightsClient && anonymousUserToken) {
-          // When `aa('init', { ... })` is called, it creates an anonymous user token in cookie.
-          // We can set it as userToken.
-          setUserTokenToSearch(anonymousUserToken);
-        }
-
-        // We consider the `userToken` coming from a `init` call to have a higher
-        // importance than the one coming from the queue.
-        if (userTokenBeforeInit) {
-          insightsClient('setUserToken', userTokenBeforeInit);
-        } else if (queuedUserToken) {
-          insightsClient('setUserToken', queuedUserToken);
-        }
-
         // This updates userToken which is set explicitly by `aa('setUserToken', userToken)`
         insightsClient('onUserTokenChange', setUserTokenToSearch, {
           immediate: true,
```

All we need to do is to make sure people use search-insights >= 1.7.2 so that we can safely remove the unnecessary code in the insights middleware.

However, we currently do not expose `version` from search-insights, so we'd better expose it first in 1.7.3 and let the insights imddleware check the version.

**Pull request closed without merging**